fix Issue 18729 - dmd -run executes in different environment

marler8997 opened the pull request:
TLDR: the following fails on windows
```
echo void main(){} > voidmain.d
dmd -run rdmd.d -m64 voidmain.d

OPTLINK (R) for Win32  Release 8.00.17
Copyright (C) Digital Mars 1989-2013  All rights reserved.
http://www.digitalmars.com/ctg/optlink.html
OPTLINK : Warning 9: Unknown Option : OUT
OPTLINK : Warning 9: Unknown Option : LIBPATH
OPTLINK : Warning 9: Unknown Option : LIBPATH
OPTLINK : Warning 9: Unknown Option : LIBPATH
...
```

I ran into this problem when I was running a D program via `dmd -run` that itself invoked DMD.  The resulting program was invoking the compiler with `-m64`, but the original `dmd -run` did not use `-m64`.  This caused a problem because the first invocation of the compiler set the `LINKCMD` environment variable to the OPTLINK executable, which caused the second invocation to use the OPTLINK executable but it treated it like the MSVC linker because it was running in 64-bit mode.

The solution here is to maintain the original environment (the environment before it was changed by the compiler) when `dmd -run` executes the compiled program.  This PR accomplishes this by saving the original values of environment variables when they are overwritten and then restoring these values before executing the compiled program.

wilzbach commented on test/runnable/sameenv.sh:
```diff
@@ -0,0 +1,10 @@
+
+$DMD -m${MODEL} -run ${EXTRA_FILES}/printenv.d > ${OUTPUT_BASE}/envFromRun.txt
+
+diff ${OUTPUT_BASE}/envFromExe.txt ${OUTPUT_BASE}/envFromRun.txt
```
I always found adding `-p` more helpful when looking at errors. Maybe we should overwrite add an `alias diff = diff -p` to the "bash runtime environment"?

wilzbach replied:
Also maybe you want to compare it with the output from `env`?

marler8997 replied:
output from env might be different, it sorts differently I believe.

wilzbach replied:
Then normalize both by  `| sort`?

marler8997 replied:
I get different outputs from `env` and this program on windows.  It sorts a bit different.  In any case, we don't really care that the output exactly matches the output of `env`, we just want to make sure running the executable directly matches `dmd -run`.

marler8997 commented on src/dmd/dinifile.d:
```diff
@@ -153,7 +153,7 @@ void updateRealEnvironment(StringTable* environment)
         memcpy(s + namelen + 1, value, valuelen);
         s[namelen + 1 + valuelen] = 0;
         //printf("envput('%s')\n", s);
-        putenv(s);
```
TODO: Not sure why this call to `putenv` exists...we should determine why this was put here.

wilzbach commented on src/dmd/link.d:
```diff
@@ -151,6 +172,37 @@ version (Posix)
+    /**
+    Call before the overriding the environment variable to save and restore it later.
+    */
+    void saveToRestoreLater()
```
How about wrapping `EnvRestore` as a voldemort type in a function and thus doing this in at construction?

marler8997 replied:
it doesn't always save the environment variable, only in some cases.

wilzbach commented on src/dmd/link.d:
```diff
@@ -151,6 +172,37 @@ version (Posix)
+struct EnvRestore(string varName)
+{
+    private string _saved;
+    private bool restore;
```
`isSaved` / `hasSaved`?

wilzbach commented on src/dmd/mars.d:
```diff
@@ -266,7 +266,8 @@ private int tryMain(size_t argc, const(char)** argv)
     getenv_setargv(readFromEnv(&environment, "DFLAGS"), &arguments);
-    updateRealEnvironment(&environment);
+    // TODO: not sure why dmd is doing this
+    //updateRealEnvironment(&environment);
```
Because it has the "great concept" of replacing `DFLAGS` completely with the content of the INI file.
See also: https://github.com/dlang/tools/pull/343

wilzbach commented on src/dmd/link.d:
```diff
@@ -32,6 +32,27 @@ version (Windows) extern (C) int putenv(const char*);
+    buffer[name.length + 1 .. fullLength ] = value[];
+    buffer[fullLength] = '\0';
+    return putenv(buffer);
+}
```
Imho `{get,put}env2` is a genuinely bad name for functions for D-ified versions.
Maybe `dgetenv` or simply use overloading?

wilzbach commented on src/dmd/link.d:
```diff
@@ -738,15 +791,16 @@ version (Windows)
-                sprintf(q, "_CMDLINE=%s", args);
-                status = putenv(q);
+                restore_CMDLINE.saveToRestoreLater();
+                status = putenv2("_CMDLINE", args);
```
Use D strings -> `args[0 .. len]`

wilzbach commented on src/dmd/link.d:
```diff
@@ -32,6 +32,27 @@ version (Windows) extern (C) int putenv(const char*);
+    auto result = getenv(name);
+    return cast(string)result[0 .. strlen(result)];
+}
+int putenv2(const char[] name, const(char)* value)
```
Do we really need the C-string overload? Remember they "must die" ;-)

ibuclaw commented on src/dmd/globals.d:
```diff
@@ -68,6 +73,75 @@ enum CPU
+/**
+Used to store an environment variable including its value.
+*/
+extern (C++) struct EnvironmentVariable
```
Why does this need to be in global? Nobody apart from dmd would use this.

ibuclaw commented on src/dmd/globals.d:
```diff
@@ -219,6 +294,7 @@ struct Param
 
     bool run; // run resulting executable
     Strings runargs; // arguments for executable
+    Array!EnvironmentVariable envToRestoreBeforeRun;
```
This is only accessed by private functions, and shouldn't be in global params accessible by all.

ibuclaw reviewed (dismissed):
For better encapsulation, these routines should be free functions and declared `extern(D)`, if dinifile or utils isn't a good place, then it's own module, where both EnvironmentVariable struct and array are private.

marler8997 commented:
@ibuclaw good suggestions.  Thanks for taking the time to review.

stefan-koch-sociomantic commented on src/win32.mak:
```diff
@@ -153,7 +153,7 @@ FRONT_SRCS=$D/access.d $D/aggregate.d $D/aliasthis.d $D/apply.d $D/argtypes.d $D
 	$D/declaration.d $D/delegatize.d $D/denum.d $D/dimport.d $D/dinifile.d $D/dinterpret.d	\
 	$D/dmacro.d $D/dmangle.d $D/dmodule.d $D/doc.d $D/dscope.d $D/dstruct.d $D/dsymbol.d $D/dsymbolsem.d		\
-	$D/lambdacomp.d $D/dtemplate.d $D/dversion.d $D/escape.d			\
+	$D/lambdacomp.d $D/dtemplate.d $D/dversion.d $D/env.d $D/escape.d			\
```
positional difference when compared to posix makefile.

wilzbach replied:
@stefan-koch-sociomantic this is fine. The windows Makefile is only semi-ordered anyhow ...

stefan-koch-sociomantic reviewed (changes requested):
Please keep the make files as much as possible in sync

marler8997 commented:
Ping

marler8997 commented:
Ping

marler8997 commented:
Re-opening, I'm hopeful that with the re-organization, that this one will be able to get in.

JinShil commented:
Forgive me if I'm missing the point, but shouldn't `dmd -m64 -run rdmd.d -m64 voidmain.d` do the right thing?  That is, the position of the switches relative to the `-run` switch determines which switches are passed to which invocation?

marler8997 commented:
Yeah that command would work. Though, it's different than the problem command that this PR fixes. Dmd uses environment variables to pass settings to other programs, but then doesn't clean up those environment variables before invoking the resulting executable with -run.

marler8997 commented:
By the way, if you're sitting in the dlang tools repo, you can do this:
```bash
./rdmd.d rdmd.d rdmd.d rdmd.d ...[repeat forver]
```
You can just keep going, and eventually it will just print out the rdmd usage...lol :)

jacob-carlborg commented on src/dmd/env.d:
```diff
@@ -0,0 +1,134 @@
+    /**
+    Access the name of the environment variable as a D string.
+    */
+    auto nameString() const { return setString[0 .. equalsIndex]; }
```
Why the `String` suffix?

jacob-carlborg commented on src/dmd/env.d:
```diff
@@ -0,0 +1,134 @@
+
+/**
+Used to store an environment variable including its value.
+*/
```
You can save two lines by using the `///` syntax.

jacob-carlborg commented on src/dmd/env.d:
```diff
@@ -0,0 +1,134 @@
+*/
+struct Env
+{
+    private const(char)* setString; /// null-terminated string of the form "VAR=VALUE"
```
Can we use a D string?

marler8997 replied:
We could but I don't see the benefit here.

marler8997 replied:
Right but this isn't being used as a string.  It's just being passed to the C function `putenv`.

marler8997 replied:
I could still make it a string, but this isn't really being used like other strings we want to convert to D strings, hence why I don't see the benefit.  Maybe you can show me a place where using a D string would prevent an error?  I don't see any.

marler8997 replied:
This member MUST be a null-terminated string.  And I've noticed that when DMD has to use a null-terminated string, it usually uses `char*` to signify that.  If I use `char[]` would that be confusing?

marler8997 replied:
Ok I've noticed other places where DMD uses D arrays for strings even when they need to be null-terminated.  So I've gone ahead and made it a `string`.

WalterBright reviewed (dismissed):
First off, bug fixes should have a bugzilla issue filed for them. Second, I'm not sure this is even a bug. Why should running a compiled program inherit the environment set by the compiler? The two have no necessary connection between them. I think it's a bug if it **did** inherit the environment variables set by the compiler. This shouldn't be polymorphism.

I.e. need a much better justification.

WalterBright commented:
Ok, I see it does have a bugzilla issue. I fixed the title to follow the usual form for bug fixes.

marler8997 commented:
> Why should running a compiled program inherit the environment set by the compiler?

It shouldn't that's the bug.

> The two have no necessary connection between them. I think it's a bug if it did inherit the environment variables set by the compiler. This shouldn't be polymorphism.

Agreed.  You're describing what the compiler is doing today.  This PR fixes it so that the executable is run from the original environment before the compiler changed it during compilation.

marler8997 commented:
To be fair, I read the description and could see how the purpose could be confused.  I updated the description to be more clear.

> The solution here is to maintain the original environment (the environment before it was changed by the compiler) when dmd -run executes the compiled program. This PR accomplishes this by saving the original values of environment variables when they are overwritten and then restoring these values before executing the compiled program.

thewilsonator commented on src/dmd/link.d:
```diff
@@ -1053,12 +1048,15 @@ version (Windows)
                     npath[5 + addpathlen] = ';';
                     memcpy(npath + 5 + addpathlen + 1, path, pathlen + 1);
-                    putenv(npath);
+                    if (0 != LocalEnvVar(cast(string)npath[0 .. length], 4).putenvRestorable)
```
please don't use the yodel style

marler8997 replied:
Actually in this cast, `putenvRestorable` is a member of `LocalEnvVar`

thewilsonator replied:
I meant remove the `0 !=`

marler8997 replied:
gotcha, done

thewilsonator commented:
Apart from the nit, this looks good.

marler8997 commented:
I emailed Walter and he just got back to me saying he's going to take another look at this PR after I clarified what it was doing.
> Thanks for clearing up my confusion. I'll see about getting to it. -Walter

marler8997 commented:
Still waiting on @WalterBright  to review.  I've emailed him a couple times and he says he hasn't had time to review yet.

In the meantime, since #9848 was merged, I've converted the new BASH test I added in this PR to D.

WalterBright commented on src/dmd/env.d:
```diff
@@ -0,0 +1,113 @@
+    return result;
+}
+
+/// Holds a `VAR=value` string that can be put into the global environment.
```
It's "name=value".

WalterBright commented on src/dmd/env.d:
```diff
@@ -0,0 +1,113 @@
+/// Holds a `VAR=value` string that can be put into the global environment.
+struct LocalEnvVar
+{
+    string nameValueCStr;        // The argument passed to `putenv`.
```
Note it's called "nameValue" here, not "VARvalue"

WalterBright replied:
Note in the comment that it has a terminating 0 past the end of the string, not as part of the name.

WalterBright commented on src/dmd/env.d:
```diff
@@ -0,0 +1,113 @@
+
+/**
+Construct a variable from `name` and `value` and put it in the environment while saving
+the current value of the environment variable.
```
No `Params:` section. Function does not construct a variable. Should return a bool.

WalterBright commented on src/dmd/env.d:
```diff
@@ -0,0 +1,113 @@
+int putenvRestorable(const(char)[] name, const(char)[] value) nothrow
+{
+    auto var = LocalEnvVar.xmalloc(name, value);
+    const result = var.putenvRestorable();
```
Having two different functions with the same name is confusing.

WalterBright commented on src/dmd/env.d:
```diff
@@ -0,0 +1,113 @@
+    Put this variable in the environment while saving the current value of the
+    environment variable.
+    Returns:
+        0 on success, non-zero on failure
```
use bool

WalterBright commented on src/dmd/env.d:
```diff
@@ -0,0 +1,113 @@
+
+    /**
+    Put this variable in the environment while saving the current value of the
+    environment variable.
```
save it where?

WalterBright commented on src/dmd/env.d:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    /**
+    Allocate a new variable via xmalloc that can be promoted to the global environment.
```
`promoted`? more like `added`

WalterBright commented on src/dmd/env.d:
```diff
@@ -0,0 +1,113 @@
+            doFree = (putenvResult != 0);
+        }
+        if (doFree)
+            mem.xfree(cast(void*)nameValueCStr.ptr);
```
get rid of `doFree` and you can collapse this to 4 lines of code instead of 7.

WalterBright commented on src/dmd/env.d:
```diff
@@ -0,0 +1,113 @@
+        foreach (var; originalVars)
+        {
+            if (0 != putenv(cast(char*)var.nameValueCStr))
+                assert(0, "putenv failed");
```
Really don't need elaborate assert's in DMD. If people can't find where an assert failed from the stack trace that gives file/line, they shouldn't be working on DMD.

WalterBright commented on src/dmd/env.d:
```diff
@@ -0,0 +1,113 @@
+    {
+        foreach (var; originalVars)
+        {
+            if (0 != putenv(cast(char*)var.nameValueCStr))
```
Not a fan of `0 !=`. We can expect some minimal knowledge of how boolean conditions work in D for people working on DMD.

WalterBright commented on src/dmd/env.d:
```diff
@@ -0,0 +1,113 @@
+    Returns:
+        a newly allocated variable that can be promoted to the global environment
+    */
+    static LocalEnvVar xmalloc(const(char)[] name, const(char)[] value) nothrow
```
Calling this `xmalloc` is confusing it with `Mem.xmalloc`. Same with `xfree`.

WalterBright commented on src/dmd/env.d:
```diff
@@ -0,0 +1,113 @@
+                return; // already saved
+        }
+        originalVars.push(LocalEnvVar.xmalloc(name,
+            name.toCStringThen!(n => getenv(n.ptr)).toDString));
```
There seems to be a lot of converting `char*` to `[]` and back again. Wouldn't it be simpler to just leave them as `char*`?

marler8997 replied:
> There seems to be a lot of converting char* to [] and back again
Alot? This is the one and only place in this change that we're calling `toCStringThen`.

We're using the `char[]` version of `name` for both the comparison and to pass to the `LocalEnvVar` allocation function.  We only need the `char*` version to call `getenv`.  I'm not sure what you'd like me to change here.

marler8997 commented:
@WalterBright I've integrated all your suggestions except the one about `toCStringThen` since I'm not sure what you want for that suggestion.  Please let me know what you'd like me to do next here.

WalterBright commented:
Thank you. In thinking about it some more, what's happening is simply storing `name=value` pairs, and avoiding making duplicate entries for `name`. This is exactly what D associative arrays do. You can dispense with the new structs, and just use:
```
__gshared string[string] envNameValues;
```
Use the `in` operator to test for membership, and `foreach` to iterate over `envNameValues`.

Commit:
Maintain same environment for dmd -run

Fix issue 18729

marler8997 commented:
Ok, I've changed the code per your suggestions.

WalterBright commented:
Much better.

marler8997 commented:
Thanks alot for seeing this through to the finish line.
